### PR TITLE
LX-1396 umount failed (part deux)

### PR DESCRIPTION
**Pull request by prakashsurya**

### CODE REVIEW

**Comment by prakashsurya** [comment]
`git-ab-pre-push` is [here](http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/78/)

File: lib/libzfs/libzfs_mount.c [hunk 1]
```diff
@@ -1434,13 +1440,17 @@ zpool_disable_datasets(zpool_handle_t *zhp, boolean_t force)
 	 */
 	for (i = 0; i < used; i++) {
 		zfs_share_proto_t *curr_proto;
+		verify(sharetab_lock() == 0);
```
Comment (grwilson) [inline lib/libzfs/libzfs_mount.c]: What about grabbing this lock outside the loop?

File: lib/libzfs/libzfs_mount.c [hunk 2]
```diff
@@ -800,6 +800,9 @@ zfs_init_libshare(libzfs_handle_t *zhandle, int service)
 {
 	int ret = SA_OK;
 
+	VERIFY(zhandle != NULL);
+	zhandle->libzfs_shareflags |= ZFSSHARE_MISS;
```
Comment (grwilson) [inline lib/libzfs/libzfs_mount.c]: We should probably just restructure this function. Effectively, the new logic is:

```
zfs_init_libhsare()
{
    zfs_unit_libshare(zhandle);
    zhandle->libzfs_sharehdl = sa_init(service);

    if (zhandle->libzfs_shardle == NULL)
        return (SA_NO_MEMORY);
    return (SA_OK);
}

Comment (prakashsurya) [reply lib/libzfs/libzfs_mount.c]: Sure. I opted for the least lines changed, to try an limit the impact on our merges.. but I can do it that way.

**Comment by prakashsurya** [comment]
Now blackbox fails with this message:
```
An error occurred during execution of 'zfs set sharenfs=off domain0/group-2/appdata_container-6/appdata_timeflow-7': [cannot unshare 'domain0/group-2/appdata_container-6/appdata_timeflow-7/archive': not found: unshare(1M) failed].
```
This is consistent with the error I see when running a local test; I'll continue to debug this new error mode and see if I can easily address it.

**Pull request closed without merging**